lib/src/bitonic_sort.dart

```diff
@@ -0,0 +1,181 @@
+    final newValA = Logic(width: 8);
+    final newValB = Logic(width: 8);
+
+    Sequential(clk, [
+      If(
+          (ascending & (_inputs[i] > _inputs[j])) |
+              (~ascending & _inputs[i].lt(_inputs[j])),
+          then: [
+            newValA < _inputs[j],
+            newValB < _inputs[i],
+          ],
+          orElse: [
+            newValA < _inputs[i],
+            newValB < _inputs[j]
+          ]),
+    ]);
+
+    _inputs[i] = newValA;
+    _inputs[j] = newValB;
+
+    for (var k = 0; k < _inputs.length; k++) {
+      _outputs.add(addOutput('y$k', width: _inputs[k].width));
+      _outputs[k] <= _inputs[k];
+    }
+  }
+}
+
+class _BitonicMerge extends Module {
+  /// A list of [Logic] that hold inputs.
+  List<Logic> _inputs = [];
+
+  /// A list of [Logic] that hold the results after [_CompareSwap].
+  final List<Logic> _outputs = [];
+
+  /// A list of [Logic] that hold the final outputs of List of result.
+  final List<Logic> _outputsFinal = [];
+
+  /// The [sorted] result.
+  List<Logic> get sorted => _outputsFinal;
+
+  /// Merge and sort [bitonicSequence] into [direction] given.
+  ///
+  /// List of [Logic] will compare and swap [Logic] position based on the
+  /// [direction] given by [BitonicSort] to first created a bitonic sequence.
+  /// The final stage will sort the bitonic sequence into sorted order
+  /// of [direction].
+  _BitonicMerge(
+      Logic clk, Logic reset, List<Logic> bitonicSequence, int direction)
+      : super(name: 'bitonic_merge') {
+    clk = addInput('clk', clk);
+    reset = addInput('reset', reset);
+
+    for (var i = 0; i < bitonicSequence.length; i++) {
+      _inputs.add(
+          addInput('x$i', bitonicSequence[i], width: bitonicSequence[i].width));
+    }
+
+    if (_inputs.length > 1) {
+      final k = _inputs.length ~/ 2;
+
+      for (var i = 0; i < 0 + k; i++) {
+        final indexA = i;
+        final indexB = i + k;
+        final swap =
+            _CompareSwap(clk, reset, _inputs, indexA, indexB, direction);
+        _inputs = swap.swapped;
+      }
+
+      for (var i = 0; i < _inputs.length; i++) {
+        _outputs.add(addOutput('y$i', width: _inputs[i].width));
+        _outputs[i] <= _inputs[i];
+      }
+
+      final mergeLeft = _BitonicMerge(clk, reset,
+          _inputs.getRange(0, _inputs.length ~/ 2).toList(), direction);
+      final mergeRight = _BitonicMerge(
+          clk,
+          reset,
+          _inputs.getRange(_inputs.length ~/ 2, _inputs.length).toList(),
+          direction);
+
+      final mergeRes = mergeLeft.sorted + mergeRight.sorted;
+
+      for (var i = 0; i < mergeRes.length; i++) {
+        _outputsFinal.add(addOutput('sorted_$i', width: mergeRes[i].width));
+        _outputsFinal[i] <= mergeRes[i];
+      }
+    } else if (_inputs.length == 1) {
+      _outputsFinal.add(addOutput('sorted_0', width: _inputs[0].width));
+      _outputsFinal[0] <= _inputs[0];
+    }
+  }
+}
+
+/// Sort [inputs] to specified order.
```

Comment:
It would be good to document the latency of the sorter (as a function of number of items sorted, probably), and also clarify that it is pipelined rather than multi-cycle.